boa_engine/src/syntax/parser/statement/declaration/hoistable/class_decl/mod.rs

```diff
@@ -0,0 +1,1073 @@
+use std::io::Read;
+
+/// Class declaration parsing.
+///
+/// More information:
+///  - [MDN documentation][mdn]
+///  - [ECMAScript specification][spec]
+///
+/// [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/class
+/// [spec]: https://tc39.es/ecma262/#prod-ClassDeclaration
+#[derive(Debug, Clone, Copy)]
+pub(super) struct ClassDeclaration {
+    allow_yield: AllowYield,
+    allow_await: AllowAwait,
+    is_default: AllowDefault,
+}
+
+impl ClassDeclaration {
+    /// Creates a new `ClassDeclaration` parser.
+    pub(super) fn new<Y, A, D>(allow_yield: Y, allow_await: A, is_default: D) -> Self
+    where
+        Y: Into<AllowYield>,
+        A: Into<AllowAwait>,
+        D: Into<AllowDefault>,
+    {
+        Self {
+            allow_yield: allow_yield.into(),
+            allow_await: allow_await.into(),
+            is_default: is_default.into(),
+        }
+    }
+}
+
+impl<R> TokenParser<R> for ClassDeclaration
+where
+    R: Read,
+{
+    type Output = Node;
+
+    fn parse(
+        self,
+        cursor: &mut Cursor<R>,
+        interner: &mut Interner,
+    ) -> Result<Self::Output, ParseError> {
+        cursor.expect(Keyword::Class, "class declaration", interner)?;
+        let strict = cursor.strict_mode();
+        cursor.set_strict_mode(true);
+
+        let token = cursor.peek(0, interner)?.ok_or(ParseError::AbruptEnd)?;
+        let name = match token.kind() {
+            TokenKind::Identifier(_) | TokenKind::Keyword(Keyword::Yield | Keyword::Await) => {
+                BindingIdentifier::new(self.allow_yield, self.allow_await)
+                    .parse(cursor, interner)?
+            }
+            _ if self.is_default.0 => Sym::DEFAULT,
+            _ => {
+                return Err(ParseError::unexpected(
+                    token.to_string(interner),
+                    token.span(),
+                    "expected class identifier",
+                ))
+            }
+        };
+        cursor.set_strict_mode(strict);
+
+        Ok(Node::ClassDecl(
+            ClassTail::new(name, self.allow_yield, self.allow_await).parse(cursor, interner)?,
+        ))
+    }
+}
+
+/// Class Tail parsing.
+///
+/// More information:
+///  - [ECMAScript specification][spec]
+///
+/// [spec]: https://tc39.es/ecma262/#prod-ClassTail
+#[derive(Debug, Clone, Copy)]
+pub(in crate::syntax::parser) struct ClassTail {
+    name: Sym,
+    allow_yield: AllowYield,
+    allow_await: AllowAwait,
+}
+
+impl ClassTail {
+    /// Creates a new `ClassTail` parser.
+    pub(in crate::syntax::parser) fn new<Y, A>(name: Sym, allow_yield: Y, allow_await: A) -> Self
+    where
+        Y: Into<AllowYield>,
+        A: Into<AllowAwait>,
+    {
+        Self {
+            name,
+            allow_yield: allow_yield.into(),
+            allow_await: allow_await.into(),
+        }
+    }
+}
+
+impl<R> TokenParser<R> for ClassTail
+where
+    R: Read,
+{
+    type Output = Class;
+
+    fn parse(
```

Comment:
This parser is way too big, and it's not clear how to follow the spec. It seems that this parser should be pretty simple: just check if the `extends` keyword exists, and if so, call the `ClassHeritage` parser, and then, expect the `{`,  call the `ClassBody` parser, then expect the `}`.

Then, we could have all the different parsers defined (well, `ClassBody` would be a type alias for `ClassElementList`, for example). Dividing this parser into multiple smaller parsers should make things more maintainable (and reusable, if any of them gets referenced anywhere else in the future).

Reply:
I refactored the parser to match the spec better. I think it got a bit better for sure. Some syntax error handling is now consolidated instead of spread throughout hundreds of lines.

The `ClassElement` parser is still ~700 lines long. We maybe can get the line count down with some helpers, but I think it is mostly the natural complexity of the parser. I guess this is probably one of the most complex parsers in spec, because we have fields, methods of all kinds, all of that in static, constructors, static blocks and all of that with private identifiers.